### PR TITLE
Server rewrite based on Tornado

### PR DESCRIPTION
**Pull request by bryevdv**
This PR is intended to be a place to have concrete discussions about technical points in the new server development. I intend to submit a new PR with all commits squashed, when this is ready to merge. 

Notes
-------
To see some basic operation with messages passed back and forth from client and server, run
```
python bk.py  # basic server
```
and 
```
python client.py # basic client
```
The class/logic in `client.py` is eventually intended to be rolled into `bokeh.session` and then real control, updating of the server will be possible with *no more ugly polling, ever*.

Concepts
--------
* `WSHandler`  &mdash; takes individual message framents, passes to `Recevier` until a message is completed, then passes message to a handler
* `Receiver` &mdash; responsible for assembling message fragments into complete messages
* `Message` &mdash; define operations on the server or client
* `Protocol` &mdash; defines a specific collection of message revisions and helps automate their creation
* `ServerSession` &mdash; an in-memory document cache for one client connection (e.g., a single browser tab)
* `ServerDocument` encapsulate a Bokeh Document on the server. Manages a work queue that serializes any storage operations on the document. 

Wire Protocol
----
Modeled loosely after the IPython protocol. A message comprises these sub-parts, sent as individual websocket messags:
```
[
    # these are required
    b'foobarbaz',       # HMAC signature
    b'{header}',        # serialized header dict
    b'{metadata}',      # serialized metadata dict
    b'{content},        # serialized content dict

    # these are optional, and come in pairs
    b'{buf_header}',    # serialized buffer header dict
    b'array'            # raw buffer payload data, possibly binary
    ...
]
```

New Document Idea
----------------

The basic idea is for `Document` to be a data structure composed of three parts: `variables`, `models`, `stores`:
```
class Document(HasProps):

    # basically a namespace for simple state to be included in Bokeh docs
    vars = Dict(String, Any)

    # like current models, except all lightweight -- no actual data
    models = Dict(String, Instance(Model))
 
    # actual data goes here
    stores = Dict(String, Instance(DataStore))
```

Various use-cases are possible: 
* sessions can copy everything when the session is created (no updates pushed to sessions)
* sessions could share/be linked to some combination of `vars`, `models`, `stores` (updates get pushed to sessions)

Additionally there should be some provision or mode for clients to *store* changes that that push to a session back to the persistent backend. Need to distinguish between changes pushed to session cache vs changes inteded to be saved. GH commit vs push analogy?

Questions
-----------
* `Message` subclasses have `handle_server` and `handle_client` methods on them, which are intended used by the server and client respectively to handle incoming messages. Would it be worthwhile to factor out the handlers? There are definitely some nice reasons to do that, but it would also spread out the codebase somewhat even more.

* The `create` method on the `Protocol`, picks out the right message and calls `create` on the message class. It just passes on any required params as `args` and `kwargs`, but the user who calls `create` has to just know what any additional params are for the specific message type. Maybe there is a better way? 

* Is there any particularly good way to start a tornado server in ipython notebook? Obviously we could just use `subprocess`, etc. but though perhaps there might be some tighter ipython integration possible.

* `Protocol` is defined in an `__init__` there's probably a better arrangement. Source layout suggestions welcome in general.  

* Currently, switching between encoded strings/bytes seems a bit ad-hoc and clunky. More consistent approach?

Todo
-----
* [ ] specify remaining messages/operations
* [ ] implement new document class
* [ ] specify storage API
* [ ] Implement storage API
 - [ ] in-memory
 - [ ] redis   
* [ ] implement message handlers
* [ ] implement JS client
* [ ] improve python client, `Session` integration
* [ ] admin page
* [ ] index page
* [ ] integrate/delegate auth checking
* [ ] update polling demos to not do polling

### CODE REVIEW

**Comment by fpliger** [comment]
My hot reaction to the proposal is absolutely positive. Need to dedicate some more time to go through it more carefully but here are a couple of quick thoughts:

* The document structure separation is a nice one. It makes things more flexible and easier to manage models/data partial changes/serialization.
* I didn't see a mention to pushes/updates (I may be overlooking). I'd strongly favor partial vars/models/data updates rather then monolitic entire updates
* With this new concept the server simply acts as a service. This means that the notion of "bokeh-server applet" is gone. Applications (either web or native) should just handle their own logic and what not on their own and "use" the bokeh-server service.. Is it correct?

Before pouring the river of thoughts I have :stuck_out_tongue_closed_eyes:  it's better to wait for the last question ..

**Comment by bryevdv** [comment]
@fpliger Yes I need to add some notes about updates, that's basically the "decided what message/operations" exist and I had planned on support both full and incremental and partial updates (i.e., a message specifically for a streaming append, for instance, that checks that all columns get appended to equally and complains informatively if not) I'll put up some initial thoughts today and we can iterate on them.

Regarding applets, I was still imagining a way to associate "scripts" or "apps" with a document, so the code could be run in the server (in thread pool executors since we are async) with "direct" access to the models. This is basically @havocp's "server callbacks" afaict.

File: bokeh/server/core/server_document.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        self._docid = docid
+        self._queue = Queue
+        self._stopped = False
+
+    def stop_worker(self):
+        self._stopped = True
+
+    @gen.coroutine
+    def workon(self, item):
+        yield self._queue.put(item)
```
Comment (pitrou) [inline bokeh/server/core/server_document.py]: This means you are not interested in the result of the processing, or even the information that the processing is done, right?

File: bokeh/server/protocol/__init__.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            metadata_json (JSON) :
+            content_json (JSON) :
+
+        Returns:
+            message
+
+        '''
+        header = json_decode(header_json)
+        return self._messages[header['msgtype']].assemble(
+            b(header_json), b(metadata_json), b(content_json)
```
Comment (pitrou) [inline bokeh/server/protocol/__init__.py]: b() should only be used for (pure ASCII) literals in Python code.
For runtime data you should have a well-defined encoding (or charset) and spell out the .encode() and .decode() calls explicitly at the protocol parsing / generation barrier.

File: bokeh/server/protocol/receiver.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    '''
+
+    def __init__(self, protocol):
+        self._protocol = protocol
+        self._expecting = HMAC
+        self._message = None
+        self._buf_header = None
+        self._failures = 0
+
+    @return_future
```
Comment (pitrou) [inline bokeh/server/protocol/receiver.py]: It seems `return_future` is Tornado's "old way" of doing event-based programming. Can't you use a `gen.coroutine` instead?

**Comment by havocp** [comment]
I'm very enthusiastic about this direction! Some comments from an initial skim of the code.

1. You can probably simplify some protocol things here by relying on TCP connection properties of a websocket.

 * the per-message HMAC shouldn't be needed unless I'm missing something - TCP already protects you from corruption. The one thing that can happen is truncation (connection closed before you get the whole message), but you could detect that with just a length instead of the expensive HMAC.
 * messages could have an incrementing integer (serial number) rather than a relatively expensive UUID, because they are "scoped" by the connection (the session or connection could have a UUID, if you need a message to have a UUID you could combine the session UUID with the message's serial number)

2. It isn't clear to me the difference between "header" and "metadata"

3. I've often seen protocol versioning schemes like this that go unused - because the easier path is generally to make compatible protocol changes, instead, like adding fields or adding new messages (plus having a convention to ignore unknown messages). But before discussing how to evolve the protocol... what is the scenario where the client and server diverge? Won't we be serving the JavaScript from the same server that speaks the protocol, so they will always be in sync?

File: bokeh/server/protocol/receiver.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+from tornado.concurrent import return_future
+
+from ..exceptions import MessageError, ProtocolError, ValidationError
+
+HMAC           = 0
+HEADER         = 1
+METADATA       = 2
+CONTENT        = 3
+BUFFER_HEADER  = 4
+BUFFER_PAYLOAD = 5
```
Comment (havocp) [inline bokeh/server/protocol/receiver.py]: A possibly-more-concise way to do this state machine is to make each state a function, and have a variable containing the current function. Like `self.current_consumer = consume_hmac` kind of thing. Then you avoid the "if" statements and the enum itself. Just a thought

**Comment by pitrou** [comment]
> Message subclasses have handle_server and handle_client methods on them, which are intended used by the server and client respectively to handle incoming messages.

IMO deciding what to do with messages is the prerogative of the server and client, respectively, so those methods have nothing to do on the message objects themselves. Message objects should provide methods pertaining to their inner logic, but that's all.

File: bokeh/server/views/ws.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        except (MessageError, ProtocolError, ValidationError) as e:
+            log.error(e)
+            # TODO (bev) : if self.receiver.failures > MAX FAILUES
+            raise gen.Return(None)
+
+        if message is None:
+            raise gen.Return(None)
+
+        # make sure the session ID from the client message matches
+        if message.header['sessid'] != self.session.id:
```
Comment (havocp) [inline bokeh/server/views/ws.py]: I would expect this check to be once when the connection opens, rather than per-message

File: bokeh/server/views/ws.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+        msg = self.session.protocol.create('ACK', self.session.id)
+        self.send_message(msg)
+
+    @gen.coroutine
+    def on_message(self, fragment):
+        ''' Process an individual wire protocol fragment.
+
+        Args:
+            fragment (UTF-8 or bytes) : wire fragment to process
```
Comment (havocp) [inline bokeh/server/views/ws.py]: this confused me for a minute. I think it's always bytes, but receiver.py may determine that the bytes are valid utf-8?

File: bokeh/server/core/server_document.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/server/core/server_document.py]: Not exactly, most of the processing here will involve updating the persistent store, perhaps as a result of a longer running computation. Those updates might trigger sending of notifications. Right now this part is just basic scaffolding, though.

File: bokeh/server/protocol/__init__.py (hunk 2 above)
Comment (bryevdv) [inline bokeh/server/protocol/__init__.py]: OK, as best I can tell these messages come in as `utf-8`

File: bokeh/server/protocol/receiver.py (hunk 3 above)
Comment (bryevdv) [inline bokeh/server/protocol/receiver.py]: Ah, maybe so I think I misunderstood, I thought `return_future` was a way of "wrapping" a regular function to return ta future but `@gen.coroutine` and changing the last line to yield should do the same as you say, I think.

File: bokeh/server/protocol/receiver.py (hunk 4 above)
Comment (bryevdv) [inline bokeh/server/protocol/receiver.py]: Yup was already thinking of making basically that change.

File: bokeh/server/views/ws.py (hunk 5 above)
Comment (bryevdv) [inline bokeh/server/views/ws.py]: Just being paranoid I guess, seemed that perhaps a client could forge a different session id if it wanted to.

File: bokeh/server/views/ws.py (hunk 6 above)
Comment (bryevdv) [inline bokeh/server/views/ws.py]: well there is a `binary` mode that is toggle-able by a flag to the ws send method. Might be misremembering but I think the API returns `utf-8

**Comment by bryevdv** [comment]
> IMO deciding what to do with messages is the prerogative of the server and client, respectively, so those methods have nothing to do on the message objects themselves. Message objects should provide methods pertaining to their inner logic, but that's all.

Yah I tend to agree, was just imagining an explosion of little handler classes. But better to do things right.

**Comment by bryevdv** [comment]
> the per-message HMAC shouldn't be needed unless I'm missing something - TCP already protects you from corruption. The one thing that can happen is truncation (connection closed before you get the whole message), but you could detect that with just a length instead of the expensive HMAC.
messages could have an incrementing integer (serial number) rather than a relatively expensive UUID, because they are "scoped" by the connection (the session or connection could have a UUID, if you need a message to have a UUID you could combine the session UUID with the message's serial number)

Full confession some of this part was cargo culting on my part. I cribbed the HMAC bit from the ipython protocol. I had assumed it was to detect corruption or adulteration of message contents over HTTP connections. I do agree that an incremental counter could suffice in conjuction with a session ID, but also with an origin tag (client or server) as well? Otherwise how to synchronize the counter across both?

> It isn't clear to me the difference between "header" and "metadata"

`metadata` is just an isolated scratch space. Imagine we want to be able to instrument messages for some kind of debug or trace mode. 

> I've often seen protocol versioning schemes like this that go unused - because the easier path is generally to make compatible protocol changes, instead, like adding fields or adding new messages (plus having a convention to ignore unknown messages). But before discussing how to evolve the protocol... what is the scenario where the client and server diverge? Won't we be serving the JavaScript from the same server that speaks the protocol, so they will always be in sync?

It's a hedge for sure, on the notion that if you *do* need it later, it's a royal pain to tack on after the fact. I have definitely seen protocol versioning used. So, I dunno. One scenario I can easily imagine is adding new message types. Which might not be incompatible per se, but it's nice to have a record of exacltly where and how these things happen. An no, I think a user could create a document e.g., load BokehJS from CDN but also expects to talk to a Bokeh server for the document data.

File: bokeh/server/core/server_document.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/server/core/server_document.py]: For instance, a computationally expensive callback gets queued here, that eventually updates the data source for a plot. That causes another chunk of work to get queues on the main IOLoop that will either push updates to the connected client, or send a message that there are updates available to pull. Does that seem reasonable? I suppose the `yield` here could return that secondary work and queue it on the other queue here, though...

Comment (bryevdv) [inline bokeh/server/core/server_document.py]: Speaking of this topic, I was planning on serializing any operations that might write to the document, and servicing reads from the document immediately.

File: bokeh/server/views/ws.py (hunk 5 above)
Comment (havocp) [inline bokeh/server/views/ws.py]: what I meant I guess is you could send a special "my session id is ___" message as the first thing on the connection and then omit it from each message, then it can't be wrong and doesn't need to be checked per-message

File: bokeh/server/views/ws.py (hunk 6 above)
Comment (havocp) [inline bokeh/server/views/ws.py]: to me it's fine if it's bytes and fine if it's utf-8 but not fine if it's unknown, if that makes sense.

File: bokeh/server/views/ws.py (hunk 5 above)
Comment (bryevdv) [inline bokeh/server/views/ws.py]: Yes sending the session ID is something that happens on when the server replies to a connection open with an ACK. I think it would still need to be sent, if we are going to simplify the message IDs?

File: bokeh/server/views/ws.py (hunk 6 above)
Comment (pitrou) [inline bokeh/server/views/ws.py]: I don't understand what "utf-8 or bytes" mean. Do you mean "unicode or bytes"? utf-8 are a particular format encoding unicode text into bytes.

**Comment by bryevdv** [comment]
OK some notes on the latest pushes: 

* separated out handler code from messages 
* plumbing for "push" and "pull" document messages
* scaffolding for storage backends and new `Document`

I think if just the new `Document` class and "push" and "pull" messages can be implemented with both an in-memory and Redis storage backed, that would be enough to reproduce the current server capability, and that additional niceties (binary array buffers, streaming and partial update messages) could be added smoothly in an incremental fashion later. I think this is the best chance for having something ready for `0.10` so I plan to try and stand up something rudimentary by the beginning of next week.

**Comment by fpliger** [comment]
> I think if just the new Document class and "push" and "pull" messages can be implemented with both an in-memory and Redis storage backed, that would be enough to reproduce the current server capability, and that additional niceties (binary array buffers, streaming and partial update messages) could be added smoothly in an incremental fashion later. I think this is the best chance for having something ready for 0.10

@bryevdv do you intend to replace the current server completely with the new (limited) one in 0.10? If so we should should out loud about those limitations so users that have bokeh-server in production (know that their app may break. I don't expect many users using bokeh-server in production but still, there may be a significant number... My gut feeling is that maybe there are users using the server streaming capabilities..

**Comment by bryevdv** [comment]
@fpliger I may not have been clear, any limitations would be relative to the future new server, not relative to the current one. As I mentioned implementing push/pull should reproduce current server capability. I don't intend to replace anything until that parity is reached. (Maybe some syntax will break or usage, but the feature set should be at least equal before replacing.)

I don't have strong feelings about shipping both servers tho I think my preference is to have a clean break. If people need the old server for a time they could just hold off on upgrading.

**Comment by bryevdv** [comment]
I also don't think getting this in `0.10` is especially likely, TBH. I guess I'd give it 50/50 odds at the moment. If not, then we can go straight to `0.11` when it is ready (definitely worth a minor version bump)

**Comment by fpliger** [comment]
@bryevdv sounds good. I misunderstood your previous comment. :+1:

File: bokeh/server/app.py [hunk 7]
```diff
@@ -1,44 +1,131 @@
...
+        try:
+            yield self._start_helper()
+        except Exception:
+            self.exit(1)
+
+    @gen.coroutine
+    def _start_helper(self):
+        self.io_loop = IOLoop.current()
+        atexit.register(self._atexit)
+        signal.signal(signal.SIGTERM, self._sigterm)
```
Comment (almarklein) [inline bokeh/server/app.py]: Probably a stupid question, but why is this done asynchronous? Can't this not simply be done in the init or start() ?

Comment (bryevdv) [inline bokeh/server/app.py]: Well, this was inspired loosely from the notebook server init code. I'll try simplifying it.

File: bokehjs/src/coffee/renderer/glyph/image_url.coffee [hunk 8]
```diff
@@ -4,35 +4,41 @@ Glyph = require "./glyph"
...
+    if not @image? or @image.length != @url.length
+      @image = (null for img in @url)
+
+    for i in [0...@url.length]
+      img = new Image()
+      img.onload = do (img, i) =>
+        return () =>
+          @image[i] = img
+          @renderer.request_render()
+      img.src = @url[i]
```
Comment (almarklein) [inline bokehjs/src/coffee/renderer/glyph/image_url.coffee]: This bit seems unrelated to this PR? Or is it just a small fix that just comes with it?

Comment (bryevdv) [inline bokehjs/src/coffee/renderer/glyph/image_url.coffee]: Came in with the ugly merge, not sure why it's showing up since it is what is in master currently AFAIK.

**Comment by bryevdv** [comment]
@pitrou thinking about this a bit more -- I think it will have to be a little more sophisticated? Basically we need to serialize all the "work" for a given document. For instance we don't want any operations that write to a given document to overlap, but operations for different documents can be concurrent. I think that it is possible for multiple operations on a single doc to overlap, with the single `ThreadPoolExecutor`?

My original thinking was that each doc would it's own executor (with thread/process limit 1, so that jobs *for that document* would get serialized). Does this seem reasonable? I guess that makes it hard to control the total number of threads/processes. Do you have any other suggestions?

**Comment by pitrou** [comment]
> BTW I played around with ProcessPoolExecutor and it seems pretty flaky compared to the thread pool executor. 

Flaky how so? It's true that process execution can be more fragile (the child process may be killed independently of the parent), although recent concurrent.futures versions should normally be quite robust.

> My original thinking was that each doc would it's own executor (with thread/process limit 1, so that jobs for that document would get serialized). Does this seem reasonable? I guess that makes it hard to control the total number of threads/processes.

How many documents do you have running tasks simultaneously? If it's a small multiple of the number of cores, then it sounds reasonable. Otherwise, you risk memory consumption issues, as well as cache thrashing. Using processes only makes the issue more severe.

Another possibility is to use a per-document Lock, but inside the main thread, not inside the workers. A threading Lock inside a worker would block the whole worker thread, while a coroutine Lock from the main thread code calling the executor will simply yield execution to other coroutines. See http://tornado.readthedocs.org/en/stable/locks.html#tornado.locks.Lock

So your scheduling code could look like:
```
    with (yield doc.lock.acquire()):
        res = yield executor.submit(doc.task, *some_args)
    raise gen.Return(res)
```

**Comment by bryevdv** [comment]
> Flaky how so? It's true that process execution can be more fragile (the child process may be killed independently of the parent), although recent concurrent.futures versions should normally be quite robust.
I mean running more than two or three clients immediately results in this:
```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/Users/bryan/anaconda/envs/tornado/lib/python3.4/threading.py", line 920, in _bootstrap_inner
    self.run()
  File "/Users/bryan/anaconda/envs/tornado/lib/python3.4/threading.py", line 868, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/bryan/anaconda/envs/tornado/lib/python3.4/concurrent/futures/process.py", line 228, in _queue_management_worker
    result_item = reader.recv()
  File "/Users/bryan/anaconda/envs/tornado/lib/python3.4/multiprocessing/connection.py", line 251, in recv
    return ForkingPickler.loads(buf.getbuffer())
EOFError
```
and then no more work is processed at all. 

Regarding number of documents, it's hard to say, I think the best we can do to start is to build in some monitoring capability so that it's more evident when adding resources is necessary.  I will play around with you second idea though, thanks!

**Comment by pitrou** [comment]
> I mean running more than two or three clients immediately results in this:

Do you have a short reproducer? If it's a Python bug, it may be useful reporting it...

**Comment by bryevdv** [comment]
> Do you have a short reproducer? If it's a Python bug, it may be useful reporting it...

I believe you can just change the thread pool to a process pool with what is currently checked in, and start up several clients at once, to reproduce.

**Comment by martindurant** [comment]
Please permit me to summarise my understanding of the architecture and flow here, so that my misunderstandings can be squashed earlier rather than later.

The whole system revolves around the BokehServer Tornado application. It receives input from Clients (which can, e.g., create and update Document) and bokehJS in users' browsers (which can request to view documents and pass back user actions in a view). It provides views and updates back to bokehJS and can request actions by clients which have created documents and are still connected. The server holds a set of Documents, which may have different states in each session (browser view). 
Messages as formatted according to a set Protocol, and the currently understood set of messages is in server.protocol.messages.* . Both clients and servers have Receivers to construct messages out of socket streams.
Clients may receive messages (e.g., callbacks from user interactions) from the server and act on them, or the simplest clients (scripts) will create a document and then exit without accepting messages. Any example script should be runnable on the server as-is.
Do JS and client communications happen on the same port and using the same Protocol, or will JS-side comms remain more similar to how they are now?

Is this on the right lines?
One question: if a Document has a distinct state for each session, and so user actions can depend upon that state, how is the state synced between the server and the client controling that Document, which has to react to actions? Or am I wrong, and the actions are processed by the server alone, in which case I don't know how the client can provide a complete description of what to do for all actions. I see that all communication happens asynchronously, but presumably they are all waiting for IO rather than doing CPU-bound processing.

**Comment by bryevdv** [comment]
@martindurant yes that is basically on the right track. One thing to clarify is that a session (containing a view of documents) is basically an in-memory cache for a browser tab.

There are complications, we are trying to figure out what to do about "callbacks" that a user might want to  specify to do actual work in response to UI events. These obviously might involve CPU work. After  a discussion with @havocp today we have a few (hopefully) clarifying ideas. I am going to try to get those in a Wiki document for wider review tonight.

**Comment by bryevdv** [comment]
Well, this sucks GH has stopped showing the complete diff. I wonder if maybe we should make one PR that moves/removes the old server, and a second PR off that one that just adds in all the new server code. @havocp thoughts? It's probably not bad idea to keep the old server around for one more release anyway.

**Comment by havocp** [comment]
It isn't clear to me which diff size limit we're hitting - it may be due to deleting the large binary files? We could try putting those back for the time being.

The entire diff is available at https://patch-diff.githubusercontent.com/raw/bokeh/bokeh/pull/2794.patch (or in a local checkout of course). The exact limits are listed at https://help.github.com/articles/what-are-the-limits-for-viewing-content-and-diffs-in-my-repository/

The hard part of keeping both servers in the source tree is the overlapping stuff that changes - state.py, _templates, Document, etc. plus of course a fair amount on the coffeescript side. It could be easier to ship from two different branches or something if we want a release with both. Or do another release with old server, then merge new server, and immediately release again, so there would be say 0.11 and 0.12 and the only difference between them is the server overhaul.

It's probably straightforward to do a PR that deletes the old server, then one that adds the new server, but of course the PR that deletes the old server would leave us in a broken state if it were merged.

**Comment by bryevdv** [comment]
Ah of course, you're right. I think we have to delete it and advise holding off on upgrades if they need the old one.

**Comment by bryevdv** [comment]
@havocp I merged in master and fixed the conflicts. I went over it pretty carefully but you might take a quick look too when you can.

**Comment by birdsarah** [comment]
small thing: 

https://github.com/bokeh/bokeh/blob/tornado/bokehjs/src/coffee/server/embed.coffee#L120

generated

    Uncaught Error: Error rendering Bokeh model: could not find tag with id: undefined

1. was cool that it picked up that i was missing a tag in my html, but in my particular case i had deliberately missed out the tag so to get a console error seemed a bit extreme - a warning would have felt more appropriate i think.
1. was disappointing it said `tag with id: undefined` would have liked it to have the actual tag `id` it was looking for.

was also confused that the source of the error was in server/embed.coffee as it was just a static plot with nothing to do with server.

**Comment by birdsarah** [comment]
two questions about selection_histogram.py example - https://github.com/bokeh/bokeh/blob/tornado/examples/plotting/server/selection_histogram.py

1. why are we doing `box_select_tool.select_every_mousemove = False` and `lasso_select_tool.select_every_mousemove = False`? I noticed it because I just changed the python default (https://github.com/bokeh/bokeh/commit/70b987dcd558363a4fd5589f10eabdd3a0fa6c1b) for box_select, so I'm guessing that we don't need the first one. It's also making me think that we need to change the python default for lasso_select.

2. Why do i not get a fresh document/session (not sure of my terminology) when I open it in another browser? if i open this example in two separate browsers, the selection state made on one is shared across the windows. Whilst this is cool, I though the default position was that fresh client instances got fresh documents. (in case it's relevant, my bokeh serve output is showing: `DEBUG:bokeh.server.tornado:[pid 5912] 3 clients connected` which is correct)

**Comment by birdsarah** [comment]
ping @havocp and @fpliger about the above as I believe you were working on this example.

**Comment by havocp** [comment]
The two browser tabs are synced because the session id is in the url giving them the same session. It has to be that way because the example works by having a Python client also talk to this same session. Remove the session ID and you should get a blank page (because no Python client to fill it in).

This whole setup is for local messing around, it isn't a production-multiuser-server-compatible setup. For a real app you would want to run the selection_histogram.py code in the server rather than as a client, which is the "bokeh serve foo.py" way. Ultimately we should port most examples to that way and only leave a couple maybe that show you can have a Python client talking to a fixed session id. But for step 1 I guess we're trying to keep the examples more similar to how they worked originally.

**Comment by havocp** [comment]
the main reason we have this single-session behavior if I remember right would be for use with a local notebook, or for something like a dashboard display or demo that runs on a single computer.
